File: typescript/typescript-sorting/selection.ts
let selectionSwaps: number = 0;
let selectionChecks: number = 0;

console.log("\nSelection Sort")

const selectionSort = (array: number[]): number[] => {

    const newArray: number[] = structuredClone(array);
    let j: number = newArray.length - 1;

    while(j > 0) {
        let biggestNumber: number = newArray[0];
        let biggestNumberLocation: number = 0;
        for(let i: number = 1; i <= j; i++) {
            selectionChecks++;
            if(newArray[i] > biggestNumber) {
                biggestNumber = newArray[i];
                biggestNumberLocation = i;
            }
        }
        const temp: number = newArray[j];
        newArray[j] = biggestNumber;
        newArray[biggestNumberLocation] = temp;
        selectionSwaps++;
        j--;
    }
    return newArray;
}

export { selectionSort, selectionSwaps, selectionChecks };

File: lua/lua-sorting/selection.lua
local deepClone = require "functions/deepClone"

local sort = {}

sort.checks = 0
sort.swaps = 0

function sort.selection(array)
    print("\nSelection Sort")

    local newArray = deepClone(array)
    local j = #newArray

    while (j > 1) do
        local biggestNumber = newArray[1]
        local biggestNumberLocation = 1
        for i = 2, j, 1 do
            sort.checks = sort.checks + 1
            if (newArray[i] > biggestNumber) then
                biggestNumber = newArray[i]
                biggestNumberLocation = i
            end
        end
        if (biggestNumberLocation ~= j) then
            local temp = newArray[j]
            newArray[j] = biggestNumber
            newArray[biggestNumberLocation] = temp
            sort.swaps = sort.swaps + 1
        end
        j = j - 1
    end
    return newArray
end

return sort
